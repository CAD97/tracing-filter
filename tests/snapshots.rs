use {
    miette::Diagnostic,
    std::{fmt, fs, path::Path},
    thiserror::Error,
    tracing_filter::simple,
};

#[derive(Debug, Error)]
#[error(transparent)]
struct StripDiagnosticUrl<'a>(&'a dyn Diagnostic);

impl Diagnostic for StripDiagnosticUrl<'_> {
    fn code<'a>(&'a self) -> Option<Box<dyn fmt::Display + 'a>> {
        self.0.code()
    }

    fn severity(&self) -> Option<miette::Severity> {
        self.0.severity()
    }

    fn help<'a>(&'a self) -> Option<Box<dyn fmt::Display + 'a>> {
        self.0.help()
    }

    fn url<'a>(&'a self) -> Option<Box<dyn fmt::Display + 'a>> {
        None
    }

    fn source_code(&self) -> Option<&dyn miette::SourceCode> {
        self.0.source_code()
    }

    fn labels(&self) -> Option<Box<dyn Iterator<Item = miette::LabeledSpan> + '_>> {
        self.0.labels()
    }

    fn related<'a>(&'a self) -> Option<Box<dyn Iterator<Item = &'a dyn miette::Diagnostic> + 'a>> {
        self.0.related().map(
            |iter| -> Box<dyn Iterator<Item = &'a dyn miette::Diagnostic> + 'a> {
                Box::new(
                    // leakage is sad but required to strip URLs from the test output [zkat/miette#136]
                    iter.map(|d| -> &dyn Diagnostic { Box::leak(Box::new(StripDiagnosticUrl(d))) }),
                )
            },
        )
    }
}

struct DisplayDiagnostic<'a>(&'a dyn Diagnostic);

impl fmt::Display for DisplayDiagnostic<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        miette::GraphicalReportHandler::new_themed(miette::GraphicalTheme::unicode_nocolor())
            .with_links(false)
            .render_report(f, &StripDiagnosticUrl(self.0))
            .or_else(|_| write!(f, "failed to pretty-print {:#?}", self.0))
    }
}

#[test]
#[cfg(feature = "regex")]
fn snapshot_simple_filter_parser() {
    fn callback(path: &Path) {
        let src = fs::read_to_string(path).unwrap();
        let (filter, report) = simple::Filter::parse(&src);

        if let Some(filter) = filter {
            insta::assert_snapshot!(Some("simple_filter"), format!("{filter:#?}"), &src);
        } else {
            insta::assert_snapshot!(Some("simple_filter"), "(compilation failed)", &src);
        }
        if let Some(report) = report {
            let report = DisplayDiagnostic(&*report);
            insta::assert_snapshot!(Some("simple_report"), format!("{report}"), &src);
        } else {
            insta::assert_snapshot!(Some("simple_report"), "(no warnings)", &src);
        }
    }

    insta::glob!("test_cases/*.env_filter", callback);
}
