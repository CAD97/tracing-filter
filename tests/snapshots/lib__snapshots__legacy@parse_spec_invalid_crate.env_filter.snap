---
source: tests/snapshots.rs
expression: "crate1::mod1=warn=info,crate2=debug"
---
Filter {
    scope: ThreadLocal { local_data: None },
    statics: DirectiveSet {
        directives: SortedSet {
            set: SortedVec {
                vec: [
                    StaticDirective {
                        target: Some(
                            "crate2",
                        ),
                        fields: [],
                        level: LevelFilter::DEBUG,
                    },
                ],
            },
        },
        level: LevelFilter::DEBUG,
    },
    dynamics: DirectiveSet {
        directives: SortedSet {
            set: SortedVec {
                vec: [],
            },
        },
        level: LevelFilter::OFF,
    },
    by_id: RwLock {
        data: {},
        poisoned: false,
        ..
    },
    by_cs: RwLock {
        data: {},
        poisoned: false,
        ..
    },
}

  × some directives had no effect

Error: 
  ⚠ 1 directives were ignored as invalid

Error: tracing_filter::legacy::InvalidLevel

  ⚠ invalid level filter specified
   ╭────
 1 │ crate1::mod1=warn=info,crate2=debug
   ·              ─────────
   ╰────
  help: valid level filters are OFF, ERROR, WARN, INFO, DEBUG, or TRACE

